File: DynamicProgramme/leetcode193.cpp/dfs.cpp
// class Solution {
// public:
//     bool wordBreak(string s, vector<string>& wordDict) {
//         unordered_map<int,bool>success;

//         const int n=s.length();
        

//         function<bool(int)>dfs=[&](int start){
//             if(start>=n){
//                 return true;
//             }

//             if(success.count(start)){
//                 return success[start];
//             }
//             for(int i=start;i<n;++i){
//                 if(find(wordDict.cbegin(),wordDict.cend(),s.substr(start,i-start+1))!=wordDict.cend() && dfs(i+1)){
//                     return success[start]=true;
//                 }
//             }
//             return success[start]=false;
//         };


//         return dfs(0);
//     }
// };

File: DynamicProgramme/leetcode193.cpp/dfs.go
// func canBreak(start int, s string, wordMap map[string]bool, memo map[int]bool) bool {

// 	if start == len(s) {

// 		return true

// 	}

// 	if res, ok := memo[start]; ok {

// 		return res

// 	}

// 	for i := start + 1; i <= len(s); i++ {

// 		prefix := s[start:i]

// 		if wordMap[prefix] && canBreak(i, s, wordMap, memo) {

// 			memo[start] = true

// 			return true

// 		}

// 	}

// 	memo[start] = false

// 	return false

// }

// func wordBreak(s string, wordDict []string) bool {

// 	wordMap := map[string]bool{}

// 	for _, v := range wordDict {

// 		wordMap[v] = true

// 	}

// 	memo := make(map[int]bool)

// 	return canBreak(0, s, wordMap, memo)

// }

